File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    rows, cols = len(maps), len(maps[0])
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([(0, 0)])
    
    while queue:
        x, y = queue.popleft()
        
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            
            if 0 <= nx < rows and 0 <= ny < cols and maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                queue.append((nx, ny))

    end = maps[rows - 1][cols - 1]
    return end if end > 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
import java.io.*;

class Solution {
    
    static int m, n;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    
    public int solution(int[][] maps) {
        m = maps.length;
        n = maps[0].length;
        
        boolean[][] visited = new boolean[m][n];      
        Queue<int[]> q = new LinkedList<>();
    
        q.add(new int[]{0, 0});
        
        while (!q.isEmpty()) {
            int[] now = q.poll();
            int x = now[0];
            int y = now[1];

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || nx >= m || ny < 0 ||  ny >= n || maps[nx][ny] == 0) {
                    continue;
                }
                
                if (visited[nx][ny]) {
                    continue;
                }
                
                maps[nx][ny] = maps[x][y] + 1;
                visited[nx][ny] = true;
                q.offer(new int[]{nx, ny});
            }
        }
        return maps[m - 1][n - 1] == 1 ? -1 : maps[m - 1][n - 1];        
    }
}